pkg/pillar/cmd/downloader/downloader.go

```diff
@@ -330,65 +343,67 @@ func handleCreate(ctx *downloaderContext, objType string,
 		log.Fatalf("handleCreate: No ObjType for %s\n",
 			config.ImageID)
 	}
-	// Start by marking with PendingAdd
-	status := types.DownloaderStatus{
-		ImageID:          config.ImageID,
-		DatastoreID:      config.DatastoreID,
-		Name:             config.Name,
-		ObjType:          objType,
-		IsContainer:      config.IsContainer,
-		RefCount:         config.RefCount,
-		LastUse:          time.Now(),
-		AllowNonFreePort: config.AllowNonFreePort,
-		PendingAdd:       true,
-	}
-	publishDownloaderStatus(ctx, &status)
-
-	// Check if we have space
-	// Update reserved space. Keep reserved until doDelete
-	// XXX RefCount -> 0 should keep it reserved.
-	kb := types.RoundupToKB(config.Size)
-	if !tryReserveSpace(ctx, &status, kb) {
-		errString := fmt.Sprintf("Would exceed remaining space. "+
-			"SizeOfAppImage: %d, RemainingSpace: %d\n",
-			kb, ctx.globalStatus.RemainingSpace)
-		log.Errorln(errString)
+	if status == nil {
+		// Start by marking with PendingAdd
+		status0 := types.DownloaderStatus{
+			ImageID:          config.ImageID,
+			DatastoreID:      config.DatastoreID,
+			Name:             config.Name,
+			ObjType:          objType,
+			IsContainer:      config.IsContainer,
+			RefCount:         config.RefCount,
+			LastUse:          time.Now(),
+			AllowNonFreePort: config.AllowNonFreePort,
+			PendingAdd:       true,
+		}
+		status = &status0
+	} else {
+		status.ImageID = config.ImageID
+		status.DatastoreID = config.DatastoreID
+		status.IsContainer = config.IsContainer
+		status.RefCount = config.RefCount
+		status.LastUse = time.Now()
+		status.Expired = false
+	}
+	publishDownloaderStatus(ctx, status)
+
+	dst, errStr := lookupDatastoreConfig(ctx, config.DatastoreID, config.Name)
+	if dst == nil {
 		status.PendingAdd = false
-		status.Size = 0
-		status.LastErr = errString
-		status.LastErrTime = time.Now()
-		status.RetryCount += 1
-		publishDownloaderStatus(ctx, &status)
-		log.Errorf("handleCreate failed for %s\n", config.Name)
+		status.SetErrorInfo(errStr)
+		publishDownloaderStatus(ctx, status)
 		return
 	}
 
 	// If RefCount == 0 then we don't yet download.
 	if config.RefCount == 0 {
 		// XXX odd to treat as error.
-		errString := fmt.Sprintf("RefCount==0; download deferred for %s\n",
+		errStr := fmt.Sprintf("RefCount==0; download deferred for %s\n",
 			config.Name)
-		log.Errorln(errString)
+		log.Errorln(errStr)
```

Comment:
This part seems to be common with maybeRetryDownload:
  1) Lookup Datastore config
  2) Reserve Memory - try download
  3) If on Error - Release memory

MAy be refactor into a separate routine:
  doDownload(config, Status ) ??

Reply:
Done.